### PR TITLE
Sigrid Benezra - Edges - Hotel

### PR DESCRIPTION
**Pull request by sdbenezra**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I originally designed the project with two classes, a Reservation Tracker to keep track of reservation lists and a Reservations class to create new reservation objects.  In the beginning I hard coded the room numbers into the Reservation Tracker class.  Eventually I felt the Reservation Tracker class seemed to do too much, so I created a Room class to hold the room numbers, and to contain the logic for finding open rooms. I also removed the code for reformatting dates, and checking the overlap of date ranges to a separate Date class.
Describe a concept that you gained more clarity on as you worked on this assignment. | The concept of separating tasks became more clear to me as I was working on formatting dates and checking for date range overlaps. I realized I could move these tasks to a separate Date class which made the Reservation Tracking class look cleaner and reduce its scope of responsibility.  Then I proceeded to create the Room class as well. Once I separated things I found it easier to make changes and fix problems.
Describe a _nominal test_ that you wrote for this assignment. | One nominal test was testing for the creation of a room reservation for a given date.
Describe an _edge case test_ that you wrote for this assignment. | I created a test to make sure an error is raised if the end date of the reservation is before the beginning date.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I did not write much pseudocode first, except near the end when I was figuring out Wave 3. I did pretty well with writing tests before code, probably 90% of the time I wrote the tests first.

### CODE REVIEW

File: lib/dates.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+require 'date'
+require 'pry'
+
+class Dates
+  attr_reader :date, :date_range
+
+  def self.date_format(date)
+    if date.class == Date
+      return date
+    else
```
Comment (droberts-sea) [inline lib/dates.rb]: This is more "parsing" (reading a string) than "formatting" (producing a string).

File: lib/dates.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    raise ArgumentError, "Invalid dates. Begin date must be at least one day prior to end date." if date_range[:begin_date] > date_range[:end_date]
+
+    return date_range
+  end
+
+
+  def self.date_range_comparison(date_range_1, date_range_2)
+    if date_range_1[:end_date] > date_range_2[:begin_date] && date_range_1[:end_date] <= date_range_2[:end_date] ||
+
+      date_range_1[:begin_date] >= date_range_2[:begin_date] && date_range_1[:begin_date] < date_range_2[:end_date] ||
```
Comment (droberts-sea) [inline lib/dates.rb]: This method name does not make it clear what you're checking for with this method. From reading the code it looks like you're checking whether they overlap, so `self.overlap?` might be a good name.

File: lib/dates.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
+require 'date'
+require 'pry'
+
+class Dates
+  attr_reader :date, :date_range
+
+  def self.date_format(date)
```
Comment (droberts-sea) [inline lib/dates.rb]: Calling this class `Dates` when Ruby has a built-in `Date` class could get confusing fast. Something more visually distinct like `DateRange` might be a better choice.

File: lib/dates.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
+require 'date'
+require 'pry'
+
+class Dates
+  attr_reader :date, :date_range
```
Comment (droberts-sea) [inline lib/dates.rb]: You have `attr_reader`s here, but I don't think you ever make instances of this class.

File: lib/room.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    available_room = find_all_open_rooms(array_of_used_rooms).first
+    return available_room
+  end
+
+  def find_block_of_rooms(num_rooms, array_of_used_rooms)
+    room_block = []
+    num_rooms.times do
+      first_available_room = find_available_room(array_of_used_rooms)
+      room_block << first_available_room
+      array_of_used_rooms << first_available_room
```
Comment (droberts-sea) [inline lib/room.rb:32]: I like the way this method builds on your previous work.

File: lib/room.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
+class Room
+  attr_reader :num_rooms, :array_of_used_rooms, :rate, :rooms
+
+  ROOMS = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
```
Comment (droberts-sea) [inline lib/room.rb:4]: I like that you extracted the logic for manipulating the list of rooms to its own class. However, I'm not convinced by the name of this class - you've called it `Room` singular, but it tracks a collection of rooms. `Rooms` or `RoomList` might be a better option.

File: lib/reservation.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+class Reservation
+  attr_reader :date_range, :room_num, :rate, :cost
+
+  def initialize(date_range, room_num, rate)
+    @date_range = {}
+    @date_range[:begin_date] = date_range[:begin_date]
+    @date_range[:end_date] = date_range[:end_date]
+    @room_num = room_num
+    @rate = rate
+    @cost = ( date_range[:end_date] - date_range[:begin_date] ) * rate
```
Comment (droberts-sea) [inline lib/reservation.rb:13]: I would make `cost` a method here rather than an instance variable. That way if the date range of this reservation changes, the cost will be updated automatically. The downside is you have to do a little more work every time you want the cost, but it's exactly the sort of work computers are really good at.

This is the sort of small thing that makes code easier to change in the future.

File: lib/reservation_tracker.rb [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+require 'pry'
+
+class ReservationTracker
+  attr_reader :rooms, :all_reservations, :cost
+
+
+  def initialize
+    @all_reservations = []
+    @rooms = Room.new
+    @all_blocks = []
```
Comment (droberts-sea) [inline lib/reservation_tracker.rb:11]: The `@rooms` variable is an opportunity to use dependency injection. That would make some of your testing a little easier - you could give it a shorter room list.

File: lib/reservation_tracker.rb [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+  def occupied_rooms(date_range)
+    occupied_rooms = []
+    date_range = Dates::date_range_format(date_range)
+    @all_reservations.each do |reservation|
+      if Dates::date_range_comparison(date_range, reservation.date_range)
+        if reservation.room_num.class == Array
+          reservation.room_num.each do |num|
+            occupied_rooms << num
+          end
```
Comment (droberts-sea) [inline lib/reservation_tracker.rb]: This is a minor example of tight coupling: the `ReservationTracker` class needs to know a lot about the internals of `Reservation` in order to extract the room numbers. A better option might be to write a `get_room_numbers` method on `Reservation`, which always returns the numbers as an array.

File: spec/dates_spec.rb [hunk 10]
```diff
@@ -0,0 +1,22 @@
+require_relative 'spec_helper'
+
+describe "Dates class" do
+  before do
+    @date = '9-10-2018'
+    @date_range = {begin_date: '9-10-2018', end_date: '10-10-2018'}
```
Comment (droberts-sea) [inline spec/dates_spec.rb]: You're missing tests for many of the methods you wrote in `Dates`. For example, your code to compare two date ranges should have a whole battery of tests here (see my note in Slack).

One of the big advantages of moving this logic into a separate class is that it allows you to test it in isolation without all the context of rooms and reservations.

File: spec/reservation_spec.rb [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+    date_range = {}
+    date_range[:begin_date] = Date.new(2018,9,1)
+    date_range[:end_date] = Date.new(2018,9,15)
+    @reservation = Reservation.new(date_range, 5, 200)
+  end
+
+  describe "Creates a new reservation"do
+    it "creates an instance of a reservation" do
+      expect(@reservation).must_be_kind_of Reservation
+    end
```
Comment (droberts-sea) [inline spec/reservation_spec.rb:16]: What about calculating the cost?

File: spec/reservation_tracker_spec.rb [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    end
+
+    it "can find an available_room for the reservation" do
+        expect (@reservation.room_num).must_equal 4
+    end
+
+    it "raises an ArgumentError if a request is made for an unavailable room" do
+      expect { 20.times do
+          @reservations.new_reservation(begin_date: '9-16-2018', end_date: '9-17-2018')
+        end }.must_raise ArgumentError
```
Comment (droberts-sea) [inline spec/reservation_tracker_spec.rb:32]: Good work on this edge case

File: spec/room_spec.rb [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    it "can list all of the rooms in the hotel" do
+      expect (@rooms.rooms).must_be_kind_of Array
+      expect (@rooms.rooms.length).must_equal 20
+    end
+  end
+
+  describe "assigning room numbers" do
+    it "assigns first available room as room_num" do
+        expect (@new_room).must_equal 3
+    end
```
Comment (droberts-sea) [inline spec/room_spec.rb:20]: What if all the rooms are unavailable?

**Comment by droberts-sea** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| 
Answer comprehension questions	| 
**Design** |
Each class is responsible for a single piece of the program | yes
Classes are loosely coupled | mostly - see inline
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | missing some - see inline
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | missing some
**Wave 3** |
Create a block of rooms | yes
Check if a block has rooms | yes
Reserve a room from a block | yes
Test coverage | missing some
**Fundamentals** | 
Names variables, classes and modules appropriately | some - see inline
Understanding of variable scope - local vs instance | yes
Can create complex logical structures utilizing variables | yes
Appropriately uses methods to break down tasks into smaller simpler tasks | yes
Understands the differences between class and instance methods | yes
Appropriately uses iterators and enumerables | yes
Appropriately writes and utilizes classes | yes
Appropriately utilizes modules as a mechanism to encapsulate functionality | no - the classes in this project should be wrapped in a `Hotel` module.
**Wrap Up** |
There is a refactors.txt file | yes
The file provides a roadmap to future changes | yes
**Additional Feedback** | 

Good work overall. This project is well-written and does a good job of meeting the project requirements. It's clear the learning goals for this assignment were met.

There are a couple of places where I see room for improvement, mostly around naming things and test coverage. I've tried to call these out inline where I saw them. However, in general I'm pretty happy with what I see. Keep up the hard work!
